### ISSUE TITLE
[sclang] [QtGUI] provide QWebSocket API in sclang

### ISSUE THREAD

**Issue by brianlheim** [supercollider/supercollider#4534]
## Motivation

See conversation at https://www.listarc.bham.ac.uk/lists/sc-users/msg65087.html

This would be useful to solve #4000, and because apparently people are already trying to do this for other projects (apparently also a quark that does it using node.js as a workaround -- https://github.com/dyfer/wsGUI.quark). So it would probably get a fair bit of use.

## Description of Proposed Feature

Add a new Qt-wrapping class in sclang that wraps QWebSocket, and wrap important components of the API.

## Plan for Implementation

For anyone interested, you could look at how other Qt classes are wrapped and exposed. It should be pretty straightforward once you're familiar with that. The WebView/QWebEngine wrapper for example also shows how to wrap code that uses callbacks.

**Comment by snappizz** [comment]
was thinking about this -- i would suggest not using QtWebSocket, just because it would be nice to have access to websockets functionality without having to build SC with Qt.

**Comment by brianlheim** [comment]
1 why and 2 what do you want to use instead?

**Comment by snappizz** [comment]
Qt is a huge dependency and users should only be forced to build SC with Qt if they need GUI features. the problem is, websockets aren't really a GUI feature, they are a network protocol. there is plenty of utility for a websocket server on a headless SC installation, particularly if you're communicating with an ordinary browser rather than a WebView.

as for what to use instead, i don't have an answer.

**Comment by brianlheim** [comment]
fair point! in that case, we could still use Qt, but just have the main application based on [QCoreApplication](https://doc.qt.io/qt-5/qcoreapplication.html#details):

> This class is used by non-GUI applications to provide their event loop.

and then link with the necessary libraries.

AFAICT the only platform where this would have an impact is BeagleBone, which I don't think has prebuilt Qt packages. I think users would need to compile the necessary Qt libraries themselves in that case. And then we don't have to expend developer and maintainer time on finding and learning a new library outside of the Qt-verse.

another viable alternative is [boost.beast](https://github.com/boostorg/beast), and there are a lot of examples in the documentation - https://www.boost.org/doc/libs/1_71_0/libs/beast/doc/html/beast/examples.html. I don't know the quality of that library but it is well-documented and Vinnie Falco (the author) seems to care a lot about maintaining it.
